File: 2023/domaci/33-navigace-skladnika/volfmatous/main.cpp
#include <iostream>
#include <vector>

const int LEFT = 0;
const int UP = 1;
const int RIGHT = 2;
const int DOWN = 3;

const int NONE = 0;
const int ROBOT = 1;
const int PILLAR = 2;
const int BOX = 3;

int rotate(int currentOrientation, bool clockwise) {
    if (clockwise) {
        return (currentOrientation + 1) % 4;
    } else {
        return (currentOrientation - 1 + 4) % 4;
    }
}

std::vector<int> moveForward(std::vector<std::vector<int>>* warehouse, int robotX, int robotY, int orientation) {
    int w = static_cast<int>(warehouse->size());
    int h = static_cast<int>((*warehouse)[0].size());

    switch (orientation) {
        case LEFT:
            for (int col = robotX - 1; col >= 0; col--) {
                if ((*warehouse)[col][robotY] == PILLAR) {
                    break;
                }

                if ((*warehouse)[col][robotY] == BOX) {
                    continue;
                }

                for (int col2 = col + 1; col2 <= robotX; col2++) {
                    (*warehouse)[col2 - 1][robotY] = (*warehouse)[col2][robotY];
                }

                (*warehouse)[robotX][robotY] = NONE;
                return {robotX - 1, robotY};
            }
            break;

        case RIGHT:
            for (int col = robotX + 1; col < w; col++) {
                if ((*warehouse)[col][robotY] == PILLAR) {
                    break;
                }

                if ((*warehouse)[col][robotY] == BOX) {
                    continue;
                }

                for (int col2 = col - 1; col2 >= robotX; col2--) {
                    (*warehouse)[col2 + 1][robotY] = (*warehouse)[col2][robotY];
                }

                (*warehouse)[robotX][robotY] = NONE;
                return {robotX + 1, robotY};
            }
            break;

        case UP:
            for (int row = robotY - 1; row >= 0; row--) {
                if ((*warehouse)[robotX][row] == PILLAR) {
                    break;
                }

                if ((*warehouse)[robotX][row] == BOX) {
                    continue;
                }

                for (int row2 = row + 1; row2 <= robotY; row2++) {
                    (*warehouse)[robotX][row2 - 1] = (*warehouse)[robotX][row2];
                }

                (*warehouse)[robotX][robotY] = NONE;
                return {robotX, robotY - 1};
            }
            break;

        case DOWN:
            for (int row = robotY + 1; row < h; row++) {
                if ((*warehouse)[robotX][row] == PILLAR) {
                    break;
                }

                if ((*warehouse)[robotX][row] == BOX) {
                    continue;
                }

                for (int row2 = row - 1; row2 >= robotY; row2--) {
                    (*warehouse)[robotX][row2 + 1] = (*warehouse)[robotX][row2];
                }

                (*warehouse)[robotX][robotY] = NONE;
                return {robotX, robotY + 1};
            }
            break;
    }

    return {robotX, robotY};
}

void printWarehouse(const std::vector<std::vector<int>>& warehouse) {
    int w = static_cast<int>(warehouse.size());
    int h = static_cast<int>(warehouse[0].size());

    for (int i = 0; i < w + 2; i++) {
        std::cout << '.';
    }

    std::cout << std::endl;

    for (int row = 0; row < h; row++) {
        std::cout << '.';

        for (int col = 0; col < w; col++) {
            char output = ' ';

            switch (warehouse[col][row]) {
                case ROBOT:
                    output = 'r';
                    break;

                case PILLAR:
                    output = 'o';
                    break;

                case BOX:
                    output = '#';
                    break;
            }

            std::cout << output;
        }

        std::cout << '.' << std::endl;
    }

    for (int i = 0; i < w + 2; i++) {
        std::cout << '.';
    }

    std::cout << std::endl;
}

int main() {
    int w, h, robotX, robotY, robotOrientation;
    std::cin >> w >> h;
    std::cin.ignore();

    std::cin >> robotX >> robotY;
    robotX -= 1;
    robotY -= 1;
    std::cin.ignore();
    robotOrientation = LEFT;

    std::vector<std::vector<int>> warehouse(w, std::vector<int>(h, 0));
    warehouse[robotX][robotY] = ROBOT;

    int p;
    std::cin >> p;
    std::cin.ignore();
    for (int i = 0; i < p; i++) {
        int x, y;
        std::cin >> x >> y;
        x -= 1;
        y -= 1;
        warehouse[x][y] = PILLAR;
        std::cin.ignore();
    }

    int b;
    std::cin >> b;
    std::cin.ignore();
    for (int i = 0; i < b; i++) {
        int x, y;
        std::cin >> x >> y;
        x -= 1;
        y -= 1;
        warehouse[x][y] = BOX;
        std::cin.ignore();
    }

    int c;
    std::cin >> c;
    std::cin.ignore();
    for (int i = 0; i < c; i++) {
        char command;
        std::cin >> command;

        switch (command) {
            case 'L':
            case 'R':
                robotOrientation = rotate(robotOrientation, command == 'R');
                break;

            case 'F':
                std::vector<int> newPosition = moveForward(&warehouse, robotX, robotY, robotOrientation);
                robotX = newPosition[0];
                robotY = newPosition[1];
                break;
        }
    }

    printWarehouse(warehouse);

    return 0;
}

File: 2023/domaci/33-navigace-skladnika/volfmatous/main.cs
using System;

internal class Program
{
    const int Left = 0;
    const int Up = 1;
    const int Right = 2;
    const int Down = 3;

    const int None = 0;
    const int Robot = 1;
    const int Pillar = 2;
    const int Box = 3;

    static void Main(string[] args)
    {
        int w = ReadInt();
        int h = ReadInt();
        Console.ReadLine();

        int robotX = ReadInt() - 1;
        int robotY = ReadInt() - 1;
        Console.ReadLine();
        int robotOrientation = Left;

        int[,] warehouse = new int[w, h];
        warehouse[robotX, robotY] = Robot;

        int p = ReadInt();
        Console.ReadLine();
        for (int i = 0; i < p; i++)
        {
            warehouse[ReadInt() - 1, ReadInt() - 1] = Pillar;
            Console.ReadLine();
        }

        int b = ReadInt();
        Console.ReadLine();
        for (int i = 0; i < b; i++)
        {
            warehouse[ReadInt() - 1, ReadInt() - 1] = Box;
            Console.ReadLine();
        }

        int c = ReadInt();
        Console.ReadLine();
        for (int i = 0; i < c; i++)
        {
            char command = char.Parse(Console.ReadLine());

            switch (command)
            {
                case 'L':
                case 'R':
                    robotOrientation = Rotate(robotOrientation, command == 'R');
                    break;

                case 'F':
                    int[] newPosition = MoveForward(warehouse, robotX, robotY, robotOrientation);
                    robotX = newPosition[0];
                    robotY = newPosition[1];
                    break;
            }
        }

        PrintWarehouse(warehouse);
    }

    static int Rotate(int currentOrientation, bool clockwise)
    {
        if (clockwise)
        {
            return (currentOrientation + 1) % 4;
        }
        else
        {
            return (currentOrientation - 1 + 4) % 4;
        }
    }

    static int[] MoveForward(int[,] warehouse, int robotX, int robotY, int orientation)
    {
        switch (orientation)
        {
            case Left:
                for (int col = robotX - 1; col >= 0; col--)
                {
                    if (warehouse[col, robotY] == Pillar)
                    {
                        break;
                    }

                    if (warehouse[col, robotY] == Box)
                    {
                        continue;
                    }

                    for (int col2 = col + 1; col2 <= robotX; col2++)
                    {
                        warehouse[col2 - 1, robotY] = warehouse[col2, robotY];
                    }

                    warehouse[robotX, robotY] = None;
                    return new int[] { robotX - 1, robotY };
                }

                break;

            case Right:
                for (int col = robotX + 1; col < warehouse.GetLength(0); col++)
                {
                    if (warehouse[col, robotY] == Pillar)
                    {
                        break;
                    }

                    if (warehouse[col, robotY] == Box)
                    {
                        continue;
                    }

                    for (int col2 = col - 1; col2 >= robotX; col2--)
                    {
                        warehouse[col2 + 1, robotY] = warehouse[col2, robotY];
                    }

                    warehouse[robotX, robotY] = None;
                    return new int[] { robotX + 1, robotY };
                }

                break;

            case Up:
                for (int row = robotY - 1; row >= 0; row--)
                {
                    if (warehouse[robotX, row] == Pillar)
                    {
                        break;
                    }

                    if (warehouse[robotX, row] == Box)
                    {
                        continue;
                    }

                    for (int row2 = row + 1; row2 <= robotY; row2++)
                    {
                        warehouse[robotX, row2 - 1] = warehouse[robotX, row2];
                    }

                    warehouse[robotX, robotY] = None;
                    return new int[] { robotX, robotY - 1 };
                }

                break;

            case Down:
                for (int row = robotY + 1; row < warehouse.GetLength(1); row++)
                {
                    if (warehouse[robotX, row] == Pillar)
                    {
                        break;
                    }

                    if (warehouse[robotX, row] == Box)
                    {
                        continue;
                    }

                    for (int row2 = row - 1; row2 >= robotY; row2--)
                    {
                        warehouse[robotX, row2 + 1] = warehouse[robotX, row2];
                    }

                    warehouse[robotX, robotY] = None;
                    return new int[] { robotX, robotY + 1 };
                }

                break;
        }

        return new int[] { robotX, robotY };
    }

    static void PrintWarehouse(int[,] warehouse)
    {
        for (int i = 0; i < warehouse.GetLength(0) + 2; i++)
        {
            Console.Write('.');
        }

        Console.WriteLine();

        for (int row = 0; row < warehouse.GetLength(1); row++)
        {
            Console.Write('.');

            for (int col = 0; col < warehouse.GetLength(0); col++)
            {
                char output = ' ';

                switch (warehouse[col, row])
                {
                    case Robot:
                        output = 'r';
                        break;

                    case Pillar:
                        output = 'o';
                        break;

                    case Box:
                        output = '#';
                        break;
                }

                Console.Write(output);
            }

            Console.WriteLine('.');
        }

        for (int i = 0; i < warehouse.GetLength(0) + 2; i++)
        {
            Console.Write('.');
        }

        Console.WriteLine();
    }

    static int ReadInt()
    {
        return int.Parse(Console.ReadLine());
    }
}
